snode/src/main/java/org/apache/rocketmq/snode/service/impl/NnodeServiceImpl.java

```diff
@@ -0,0 +1,218 @@
+                } catch (Exception ex) {
+                    log.warn("Register Snode to Nnode addr: {} error, ex:{} ", nodeAddress, ex);
+                }
+            }
+        } else {
+            log.warn("Nnode server list is null");
+            throw new RemotingSendRequestException("Nnode server list is null");
+        }
+    }
+
+    @Override
+    public void updateTopicRouteDataByTopic() {
+        Set<String> topSet = topicRouteDataMap.keySet();
+        for (String topic : topSet) {
+            try {
+                TopicRouteData topicRouteData = getTopicRouteDataByTopic(topic, false);
+                topicRouteDataMap.put(topic, topicRouteData);
+            } catch (Exception ex) {
+                log.error("Update topic {} error: {}", topic, ex);
+            }
+        }
+    }
+
+    private TopicRouteData getTopicRouteDataByTopicFromNnode(String topic,
+        boolean allowTopicNotExist) throws MQClientException, InterruptedException, RemotingTimeoutException, RemotingSendRequestException, RemotingConnectException {
+        GetRouteInfoRequestHeader requestHeader = new GetRouteInfoRequestHeader();
+        requestHeader.setTopic(topic);
+
+        RemotingCommand request = RemotingCommand.createRequestCommand(RequestCode.GET_ROUTEINTO_BY_TOPIC, requestHeader);
+        RemotingCommand response = this.snodeController.getRemotingClient().invokeSync(null, request, SnodeConstant.DEFAULT_TIMEOUT_MILLS);
+        log.info("GetTopicRouteInfoFromNameServer response: " + response);
+        assert response != null;
+        switch (response.getCode()) {
+            case ResponseCode.TOPIC_NOT_EXIST: {
+                if (allowTopicNotExist && !topic.equals(MixAll.AUTO_CREATE_TOPIC_KEY_TOPIC)) {
+                    log.warn("Topic [{}] RouteInfo is not exist value", topic);
+                }
+                break;
+            }
+            case ResponseCode.SUCCESS: {
+                byte[] body = response.getBody();
+                if (body != null) {
+                    return TopicRouteData.decode(body, TopicRouteData.class);
+                }
+            }
+            default:
+                break;
+        }
+
+        throw new MQClientException(response.getCode(), response.getRemark());
+    }
+
+    @Override
+    public TopicRouteData getTopicRouteDataByTopic(
+        String topic,
+        boolean allowTopicNotExist) throws MQClientException, InterruptedException, RemotingTimeoutException, RemotingSendRequestException, RemotingConnectException {
+        if (topic == null || "".equals(topic)) {
+            return null;
+        }
+
+        TopicRouteData topicRouteData = topicRouteDataMap.get(topic);
+        if (topicRouteData == null) {
+            topicRouteData = getTopicRouteDataByTopicFromNnode(topic, allowTopicNotExist);
+            if (topicRouteData != null) {
+                topicRouteDataMap.put(topic, topicRouteData);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **135, 136, and 139**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread's `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **142**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.